dali/pipeline/data/tensor_vector.cc

```diff
@@ -12,268 +12,494 @@
+                        const TypeInfo &type_info, AccessOrder order = {},
+                        bool use_copy_kernel = false) {
+  auto num_samples = src.num_samples();
+  SmallVector<const void *, 256> srcs;
+  srcs.reserve(num_samples);
+  SmallVector<void *, 256> dsts;
+  dsts.reserve(num_samples);
+  SmallVector<Index, 256> sizes;
+  sizes.reserve(num_samples);
+  for (int i = 0; i < num_samples; i++) {
+    dsts.emplace_back(dst.raw_mutable_tensor(i));
+    srcs.emplace_back(src.raw_tensor(i));
+    sizes.emplace_back(src.shape()[i].num_elements());
+  }
+
+  type_info.template Copy<SrcBackend, DstBackend>(dsts.data(), srcs.data(), sizes.data(),
+                                                  num_samples, order.stream(), use_copy_kernel);
+}
+
+
+template <typename DstBackend, typename SrcBackend, template <typename> typename DstBatch>
+void CopySamplewiseImpl(DstBatch<DstBackend> &dst, const void *src, const TypeInfo &type_info,
+                        AccessOrder order = {}, bool use_copy_kernel = false) {
+  auto num_samples = dst.num_samples();
+  SmallVector<void *, 256> dsts;
+  dsts.reserve(num_samples);
+  SmallVector<Index, 256> sizes;
+  sizes.reserve(num_samples);
+  for (int i = 0; i < num_samples; i++) {
+    dsts.emplace_back(dst.raw_mutable_tensor(i));
+    sizes.emplace_back(dst.shape()[i].num_elements());
+  }
+
+  type_info.template Copy<DstBackend, SrcBackend>(dsts.data(), src, sizes.data(), num_samples,
+                                                  order.stream(), use_copy_kernel);
+}
+
+
+template <typename DstBackend, typename SrcBackend, template <typename> typename SrcBatch>
+void CopySamplewiseImpl(void *dst, const SrcBatch<SrcBackend> &src, const TypeInfo &type_info,
+                        AccessOrder order = {}, bool use_copy_kernel = false) {
+  auto num_samples = src.num_samples();
+  SmallVector<const void *, 256> srcs;
+  srcs.reserve(num_samples);
+  SmallVector<Index, 256> sizes;
+  sizes.reserve(num_samples);
+  for (int i = 0; i < num_samples; i++) {
+    srcs.emplace_back(src.raw_tensor(i));
+    sizes.emplace_back(src.shape()[i].num_elements());
+  }
+
+  type_info.template Copy<DstBackend, SrcBackend>(dst, srcs.data(), sizes.data(), num_samples,
+                                                  order.stream(), use_copy_kernel);
+}
+
+/** @} */  // end of copy_impl
+
+}  // namespace
+
+
 template <typename Backend>
-TensorVector<Backend>::TensorVector()
-    : views_count_(0), curr_num_tensors_(0), tl_(std::make_shared<TensorList<Backend>>()) {}
+TensorVector<Backend>::TensorVector() : curr_num_tensors_(0) {}
 
 
 template <typename Backend>
-TensorVector<Backend>::TensorVector(int batch_size)
-    : views_count_(0),
-      curr_num_tensors_(0),
-      tl_(std::make_shared<TensorList<Backend>>(batch_size)) {
+TensorVector<Backend>::TensorVector(int batch_size) : curr_num_tensors_(0) {
+  // This is why we can't have nice things as `dim = 0` is already occupied
```

Comment:
That's hardly the reason. The real reason we "can't have nice things", as you call it, is the very fact that there can be such a thing as an empty Tensor(List) - [i]at all[/i].